File: internal/ipc/log_message.go
package ipc

import (
	"time"
)

type LogLevel string

const (
	LogLevelDebug   LogLevel = "DEBUG"
	LogLevelInfo    LogLevel = "INFO"
	LogLevelWarning LogLevel = "WARNING"
	LogLevelError   LogLevel = "ERROR"
	LogLevelFatal   LogLevel = "FATAL"
)

type LogMessage struct {
	Level        LogLevel `json:"level"`
	Message      string   `json:"message"`
	Timestamp    string   `json:"timestamp"`
	SourceFile   *string  `json:"source_file"`
	FunctionName *string  `json:"function_name"`
	LineNumber   *int32   `json:"line_number"`
	HTMLClass    *string  `json:"html_class"`
}

func NewLogMessage(
	level LogLevel,
	message string,
) LogMessage {
	return LogMessage{
		Level:        level,
		Message:      message,
		Timestamp:    time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
		SourceFile:   nil,
		FunctionName: nil,
		LineNumber:   nil,
		HTMLClass:    nil,
	}
}

File: python/adb_auto_player/ipc/log_message.py
"""ADB Auto Player Logging Module."""

from datetime import datetime, timezone

from pydantic import BaseModel


class LogLevel:
    """Logging levels."""

    TRACE = "TRACE"
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    FATAL = "FATAL"


class LogMessage(BaseModel):
    """Log message class."""

    level: str
    message: str
    timestamp: datetime
    source_file: str | None = None
    function_name: str | None = None
    line_number: int | None = None
    html_class: str | None = None

    def to_dict(self):
        """Convert LogMessage to dictionary for JSON serialization."""
        return {
            "level": self.level,
            "message": self.message,
            "timestamp": datetime.now(tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[
                :-3
            ]
            + "Z",
            "source_file": self.source_file,
            "function_name": self.function_name,
            "line_number": self.line_number,
            "html_class": self.html_class,
        }
